config-linux.md

```diff
@@ -466,19 +466,26 @@ The following parameters can be specified to set up the controller:
 The following parameters can be specified for the container:
 
 * **`l3CacheSchema`** *(string, OPTIONAL)* - specifies the schema for L3 cache id and capacity bitmask (CBM).
-    If `l3CacheSchema` is set, runtimes MUST write the value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.
+* **`memBwSchema`** *(string, OPTIONAL)* - specifies the schema of memory bandwidth percentage per L3 cache id.
```

Comment:
Can we get a line after this like:

> The value MUST start with `MB:` and MUST NOT contain newlines.

I'd like a similar line with MUST -> SHOULD for `l3CacheSchema` (or also with MUST if maintainers are comfortable with the [compat impact][1]).

Then down below, I'd like two paragraphs like:

> If both `l3CacheSchema` and `memBwSchema` are set, runtimes MUST write the combined value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.  If `l3CacheSchema` contains a line beginning with `MB:`, the value written to `schemata` MUST be the non-`MB:` line(s) from `l3CacheSchema` and the line from `memBWSchema`.
>
> If either `l3CacheSchema` or `memBwSchema` is set, runtimes MUST write the value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.
> 

[1]: https://github.com/opencontainers/runtime-spec/pull/932#discussion_r145822155

Reply:
@wking 
> The value MUST start with MB: and MUST NOT contain newlines.

Yes, that is right. I will add this line for `memBwSchema`.

> I'd like a similar line with MUST -> SHOULD for l3CacheSchema (or also with MUST if maintainers are comfortable with the compat impact).

Make sense to me. I will add this line for `l3CacheSchema`.

> If l3CacheSchema contains a line beginning with MB:, the value written to schemata MUST be the non-MB: line(s) from l3CacheSchema and the line from memBWSchema.

This is not the case. `l3CacheSchema` will never contains a line beginning with `MB:`. It only contains a line beginning with `L3:` and `memBWSchema` only contains a line beginning with `MB:`. If both `l3CacheSchema` and `memBwSchema` are set, runc runtimes will handle this case in Set() to write the combined value: "`l3CacheSchema`"+"\n"+ "`memBwSchema`", e.g., "L3:0=7f0;1=1f\nMB:0=20;1=70" to `schemata` file.

The paragraphs may like:
```
* **`l3CacheSchema`** *(string, OPTIONAL)* - specifies the schema for L3 cache id and capacity bitmask (CBM).
    The value MUST start with `L3:` and MUST NOT contain newlines.
* **`memBwSchema`** *(string, OPTIONAL)* - specifies the schema of memory bandwidth percentage per L3 cache id.
    The value MUST start with `MB:` and MUST NOT contain newlines.

    If both `l3CacheSchema` and `memBwSchema` are set, runtimes MUST write the combined value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.
    If either `l3CacheSchema` or `memBwSchema` is set, runtimes MUST write the value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.
```

Reply:
> > If `l3CacheSchema` contains a line beginning with `MB:`, the value written to schemata MUST be the non-`MB:` line(s) from `l3CacheSchema` and the line from `memBWSchema`.
>
> This is not the case. `l3CacheSchema` will never contains a line beginning with `MB:`. It only contains a line beginning with `L3:`...

That would have been nice, but I see no such MUST in 1.0.0, which means it may be too late to add it now.  To add it before 2.0 (where we intend to drop `l3CacheSchema` anyway), you'd need to make a "nobody was actually doing that, despite it being technically legal" argument.  And withought an exhaustive list of config authors and a relatively straightforward compat spec, it may not be worth trying to build that case.

Reply:
@wking 
> That would have been nice, but I see no such MUST in 1.0.0, which means it may be too late to add it now. 

How about removing these two lines?
`- The value MUST start with "L3:" and MUST NOT contain newlines.`
`- The value MUST start with "MB:" and MUST NOT contain newlines.`

In my opinion, it looks like runtime implementation detail. The validity of `l3CacheSchema` and `memBwSchema` will be checked in runc, and the validity of string writing to schemata file will be also checked by kernel.

Reply:
> In my opinion, it looks like runtime implementation detail. The validity of `l3CacheSchema` and `memBwSchema` will be checked in runc…

I think all configs not explicitly denied by the spec are allowed.  The 1.0.0 spec contains no `L3:` prefix requirement for `l3CacheSchema`, and it contains no anti-newline requirement for `l3CacheSchema`.  And runc [currently writes the opaque `l3CacheSchema` value to the `schemata` file][1].  So folks writing 1.0.0 configs can *already* set `l3CacheSchema` to `L3:0=7f0;1=1f\nMB:0=20;1=70` or `MB:0=20;1=70` and have a compliant config which is supported by runc.  I don't want to break those users, which means we need to either:

a. [Make a case for those users not existing][2], or
b. Add spec wording to [SHOULD against those configs, and MUST the runtime procedure for handling them if/when they occur][3].

b seems much more straightforward to me.

If we do not do either and those users happen to exist, having runc start erroring for `l3CacheSchema` like `L3:0=7f0;1=1f\nMB:0=20;1=70` is just pulling the rug out from under those users in order to satisfy an aesthetic concern about consistency between the existing `l3CacheSchema` and the new `memBwSchema`.  I think there is a benefit to consistency (it makes getting acquainted with the spec easier for newcomers), but it's not worth breaking existing users to achieve it. 

[1]: https://github.com/opencontainers/runc/blob/9a1186d128c4c4df4b2c6722976b49898cbe2e03/libcontainer/intelrdt/intelrdt.go#L510-L513
[2]: https://github.com/opencontainers/runtime-spec/pull/932#discussion_r146109506
[3]: https://github.com/opencontainers/runtime-spec/pull/932#discussion_r145983198

Reply:
@wking 
> a. Make a case for those users not existing, or
b. Add spec wording to SHOULD against those configs, and MUST the runtime procedure for handling them if/when they occur.

> b seems much more straightforward to me.

I'd like to replace wording MUST with SHOULD for `l3CacheSchema` and new `memBwSchema`.

` The value SHOULD start with "L3:" and SHOULD NOT contain newlines.`
` The value SHOULD start with "MB:" and SHOULD NOT contain newlines.`

Reply:
> I'd like to replace wording MUST with SHOULD for `l3CacheSchema` and new `memBwSchema`.

I [prefer MUSTs for `memBwSchema`][1]; why would you SHOULD there?  It's a new property, so we do not have `l3CacheSchema`'s backwards-compat constraints.

And either way I think you need something like [the paragraphs I recommended earlier][1] to require runtimes to handle `l3CacheSchema` SHOULD violations in a consistent way.

[1]: https://github.com/opencontainers/runtime-spec/pull/932#discussion_r145983198

Reply:
@wking 

Thank you for your scrupulous review and good suggestions.
I have updated this PR to address the comments.

The documentation for `IntelRdt` may look like:

## <a name="configLinuxIntelRdt" />IntelRdt

**`intelRdt`** (object, OPTIONAL) represents the [Intel Resource Director Technology][intel-rdt-cat-kernel-interface].
    If `intelRdt` is set, the runtime MUST write the container process ID to the `<container-id>/tasks` file in a mounted `resctrl` pseudo-filesystem, using the container ID from [`start`](runtime.md#start) and creating the `<container-id>` directory if necessary.
    If no mounted `resctrl` pseudo-filesystem is available in the [runtime mount namespace](glossary.md#runtime-namespace), the runtime MUST [generate an error](runtime.md#errors).
    If `intelRdt` is not set, the runtime MUST NOT manipulate any `resctrl` pseudo-filesystems.

The following parameters can be specified for the container:

* **`l3CacheSchema`** *(string, OPTIONAL)* - specifies the schema for L3 cache id and capacity bitmask (CBM).
    The value SHOULD start with `L3:` and SHOULD NOT contain newlines.
* **`memBwSchema`** *(string, OPTIONAL)* - specifies the schema of memory bandwidth percentage per L3 cache id.
    The value MUST start with `MB:` and MUST NOT contain newlines.

    If both `l3CacheSchema` and `memBwSchema` are set, runtimes MUST write the combined value to the `schemata` file in the <container-id> directory discussed in intelRdt. If `l3CacheSchema` contains a line beginning with `MB:`, the value written to `schemata` file MUST be the non-`MB:` line(s) from `l3CacheSchema` and the line from `memBWSchema`.

    If either `l3CacheSchema` or `memBwSchema` is set, runtimes MUST write the value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.

    If neither `l3CacheSchema` nor `memBwSchema` is set, runtimes MUST NOT write to `schemata` files in any `resctrl` pseudo-filesystems.

### Example

Consider a two-socket machine with two L3 caches where the default CBM is 0x7ff and the max CBM length is 11 bits,
and minimum memory bandwidth of 10% with a memory bandwidth granularity of 10%.

Tasks inside the container only have access to the "upper" 7/11 of L3 cache on socket 0 and the "lower" 5/11 L3 cache on socket 1,
and may use a maximum memory bandwidth of 20% on socket 0 and 70% on socket 1.

```json
"linux": {
    "intelRdt": {
        "l3CacheSchema": "L3:0=7f0;1=1f",
        "memBwSchema": "MB:0=20;1=70"
    }
}
```